woudc_data_registry/models.py

```diff
@@ -325,17 +332,27 @@ class Instrument(base):
 
     def __init__(self, dict_):
         self.station_id = dict_['station_id']
-        self.dataset_name = dict_['dataset_name']
-        self.dataset_level = str(float(dict_['dataset_level']))
         self.dataset_form = str(dict_.get('dataset_form', '1'))  # noqa; defaults to '1' if empty/none/no Key found
-        self.contributor = dict_['contributor']
-        self.project = dict_['project']
 
         self.name = dict_['name']
         self.model = dict_['model']
         self.serial = dict_['serial']
 
-        self.generate_ids()
+        try:
+            self.instrument_id = dict_['instrument_id']
+            self.dataset_id = dict_['dataset_id']
+            self.deployment_id = dict_['deployment_id']
+        except KeyError:
+            LOGGER.warning(
+                'deployment_id not found in dict, generating from '
+                'station_id, contributor and project'
+            )
+            self.dataset_name = dict_['dataset_name']
+            self.dataset_level = str(float(dict_['dataset_level']))
+            self.contributor = dict_['contributor']
+            self.project = dict_['project']
+            self.generate_ids()
```

Comment:
I wonder if `generate_ids()` can be executed outside the `except` clause? So it always gets executed when the class is instantiated.
- if so, perhaps just modify `generate_ids()` itself to ignore generating it if it's not empty (`instrument_id`, `dataset_id`, `deployment_id` is not empty separately)